### ISSUE TITLE
Figure out if SerializingExecutor should have any interruption logic

### ISSUE THREAD

**Issue by cpovirk** [google/guava#2440]
If someone submits two `FutureTask` instances and calls `cancel(true)` on the first, that interrupt will propagate from the first task to the second. Perhaps worse, it will propagate *only if* the second task is submitted before the first is done.

Possibly we should do something about this. If so, the right thing to do might be to check for interruption each iteration. If the thread was interrupted, then exit from the current `QueueWorker.run`, calling `startQueueWorker` to queue up another. In the scenario described above, this will start a new, un-interrupted worker. In the other possible scenario, that of a `shutdownNow` call on the delegate executor, the `execute` call will fail (and hopefully the exception will at least be dumped to standard error). That's probably a defensible thing to do, since it's similar to what would have happened to other un-executed tasks (except that those would be returned from `shutdownNow`... though in practice, I think almost no one looks at those, anyway).

But... I think there's a good case to be made that submitting a `FutureTask` and calling `cancel(true)` on it is a bogus thing to do: [We shouldn't be interrupting threads that we don't own.](https://github.com/google/guava/issues/1989#issuecomment-199262664) So maybe there's nothing that `SerializingExecutor` needs to do here.

**Comment by charlesmunger** [comment]
Leaking interrupt status from one task to another is a really unfortunate behavior. And while I don't disagree with your statement "we shouldn't be interrupting threads we don't own", I do think that something being the delegate of a SerializingExecutor is transferring ownership in a sense.

**Comment by cpovirk** [comment]
I do think we can at least make a case that `SerializingExecutor` takes ownership of the underlying executor's threads. But I think that only *partially* lets us wriggle out of this one: If `SerializingExecutor` is the thread owner, and if we can assume that the underlying executor implements cancellation "normally," then yes, we can reasonably allow it to clear the interrupt. But we still can't reasonably allow some *user* of `SerializingExecutor` to interrupt the thread. Well, I guess we could, assuming that we document all the caveats. It would be a little hacky, but it would probably work.

Alternatively, we could do is make `SerializingExecutor` an `ListeningExecutorService` and then enable `submit()` so that users can get a `Future` back that they can then `cancel(true)`. Then `SerializingExecutor` is still in control of the interruption.

**Comment by vharuty** [comment]
@cpovirk Hi, please let me know what is the status of this issue, cause I would like to cotribute.

**Comment by cpovirk** [comment]
This is another case in which I'm not yet sure what we want to do :(

**Comment by cpovirk** [comment]
@vharuty , if you want a medium-sized project, I suggest https://github.com/google/guava/issues/1190#issuecomment-61350932

The specific comment that I've linked to is about reimplementing `Splitter.onPattern` to follow the ECMAScript contract. I can't guarantee that we would take the result, but I would love to have the implementation to test against our codebase to see if it's practical to use.

**Comment by vharuty** [comment]
@cpovirk Hi, did I understand properly, do I need to re-implement the ```Splitter.onPattern``` function?
BTW yesterday I did a code-change which caused to return ['o','o'] for input data as {input: "foo"; regex: "(?=o)|(?<=o)"}, may I ask for a review? (before it was returning just ['0']).
Another question may be I should create another "Issue" in order to track everything in right place?

**Comment by cpovirk** [comment]
Right, you'd be changing `Splitter.onPattern`. You might need to change other code, too.

I can look at the code change you made, sure. You can post directly on #1190.
